WalletWasabi/WabiSabi/Client/CredentialDependencies/CredentialEdgeSet.cs

```diff
@@ -0,0 +1,253 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public record CredentialEdgeSet
+	{
+		public CredentialType CredentialType { get; init; }
+		public ImmutableDictionary<RequestNode, ImmutableHashSet<CredentialDependency>> Predecessors { get; init; } = ImmutableDictionary.Create<RequestNode, ImmutableHashSet<CredentialDependency>>();
+		public ImmutableDictionary<RequestNode, ImmutableHashSet<CredentialDependency>> Successors { get; init; } = ImmutableDictionary.Create<RequestNode, ImmutableHashSet<CredentialDependency>>();
+		public ImmutableDictionary<RequestNode, long> EdgeBalances { get; init; } = ImmutableDictionary.Create<RequestNode, long>();
+
+		public long Balance(RequestNode node) => node.InitialBalance(CredentialType) + EdgeBalances[node];
+
+		public ImmutableHashSet<CredentialDependency> InEdges(RequestNode node) => Predecessors[node];
+
+		public ImmutableHashSet<CredentialDependency> OutEdges(RequestNode node) => Successors[node];
+
+		public int InDegree(RequestNode node) => InEdges(node).Count;
+
+		public int OutDegree(RequestNode node) => OutEdges(node).Where(x => x.Value != 0).Count();
+
+		public int ZeroOnlyOutDegree(RequestNode node) => OutEdges(node).Where(x => x.Value == 0).Count();
```

Comment:
I would prefer conditional counting:

```suggestion
		public int OutDegree(RequestNode node) => OutEdges(node).Count(x => x.Value != 0);

		public int ZeroOnlyOutDegree(RequestNode node) => OutEdges(node).Count(x => x.Value == 0);
```

Reply:
It will do it in the same way and the difference is only in readability.